### PR TITLE
dmd 2.071.0

### PR DESCRIPTION
**Pull request by John-Colvin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-core/blob/master/.github/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Does your submission pass `brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting)?
- [x] Have you built your formula locally prior to submission with `brew install <formula>`?

### CODE REVIEW

**Comment by apjanke** [comment]
Build is failing with linker not finding some libraries. Maybe this formula is missing some dependencies?

```
==> make install DMD=/usr/local/Cellar/dmd/2.071.0/bin/dmd INSTALL_DIR=/usr/local/Cellar/dmd/2.071.0 MODEL=64 -f posix.mak
/usr/local/Cellar/dmd/2.071.0/bin/dmd -m64 -w -ofgenerated/osx/64/rdmd rdmd.d
/usr/local/Cellar/dmd/2.071.0/bin/dmd -m64 -w -ofgenerated/osx/64/ddemangle ddemangle.d
/usr/local/Cellar/dmd/2.071.0/bin/dmd -m64 -w -ofgenerated/osx/64/catdoc catdoc.d
/usr/local/Cellar/dmd/2.071.0/bin/dmd -m64 -w -ofgenerated/osx/64/detab detab.d
ddemangle.d(17): Deprecation: module std.c.stdlib is deprecated - Import core.stdc.stdlib or core.sys.posix.stdlib instead
ld: library not found for -lphobos2
clang: error: linker command failed with exit code 1 (use -v to see invocation)
--- errorlevel 1
make: *** [generated/osx/64/detab] Error 1
make: *** Waiting for unfinished jobs....
ld: library not found for -lphobos2
clang: error: linker command failed with exit code 1 (use -v to see invocation)
--- errorlevel 1
make: *** [generated/osx/64/catdoc] Error 1
ld: library not found for -lphobos2
clang: error: linker command failed with exit code 1 (use -v to see invocation)
--- errorlevel 1
make: *** [generated/osx/64/rdmd] Error 1
ld: library not found for -lphobos2
clang: error: linker command failed with exit code 1 (use -v to see invocation)
--- errorlevel 1
make: *** [generated/osx/64/ddemangle] Error 1
==> Formula
Path: /usr/local/Library/Taps/homebrew/homebrew-core/Formula/dmd.rb
```

**Comment by John-Colvin** [comment]
this is very odd. `phobos` is installed as part of dmd and is definitely done before this. Just checked on two different macs: installing, reinstalling and upgrading dmd worked fine.

**Comment by John-Colvin** [comment]
has anything change in homebrew that would break
```
     etc.install conf
     install_new_dmd_conf
```
(see `dmd.rb` for definition of `install_new_dmd_conf`)

**Comment by John-Colvin** [comment]
I don't understand this. It works perfectly on my machines and the config files in the test output look fine...

**Comment by mikemcquaid** [comment]
We've changed the behaviour of `etc.install` of directories so they behave like `etc.install` of files, that may be it. Also, `install_new_dmd_conf` should not be installing the new one every time but instead be `grep`ing for something actually incompatible to avoid moving perfectly valid configuration on every reinstall (as I read it, I could be wrong)

**Comment by apjanke** [comment]
Weird. I'll try to reproduce, too, then.

Those files in etc could be a difference: even if the "etc.install dir" behavior change doesn't effect this, the `brew test-bot` process does `--cleanup`, which I believe results in a clean `$HOMEBREW_PREFIX/etc` directory at the beginning of each installation. But for normal users, the `etc.install` behavior is to leave those files behind upon uninstall. So you'd be testing against an already-populated `etc`.

Agreed with Mike on just grepping for incompatible lines. In this case, updating `DFLAGS=-I#{include}/dlang/dmd -L-L#{lib}` to point to the new one. The whole point of leaving `etc` files alone is to preserve any user-supplied changes.

Maybe `install_new_dmd_conf` could be avoided entirely, though, by using the `opt` prefix to `dmd` instead of `#{include} and #{lib}`. `opt` is a link to the currently-installed version, without the version name in the path, so it doesn't change upon upgrades, and you wouldn't need to change the file. Would simplify things, because editing config files in-place sucks and is difficult to get correct.

File: Formula/dmd.rb [hunk 1]
```diff
@@ -98,6 +87,9 @@ def install
     cp_r ["phobos/std", "phobos/etc"], include/"dlang/dmd"
     lib.install Dir["druntime/lib/*", "phobos/**/libphobos2.a"]
 
+    system "which", "dmd"
```
Comment (apjanke) [inline Formula/dmd.rb]: Please remove the `which dmd` and `dmd --help` from the `install` method. They're not necessary, and they could cause it to error, depending on how strict your environment is and whether you have a previous `dmd` install. (i.e. the `brew link` step is what populates `/usr/local/bin`, and it takes place after `install()` returns, so if this is your first time installing `dmd`, `which dmd` is expected to fail at this point.)

**Comment by apjanke** [comment]
Okay, I can reproduce this locally, but only when running it under `brew test-bot`, not a normal `brew install -s dmd`.

To reproduce this yourself:
* check out the branch for this PR
* `mkdir -p ~/tmp/test-bot; cd ~/tmp/test-bot`
* brew test-bot --ci-pr --skip-homebrew --skip-setup dmd

This is on 10.9.5 Mavericks. The error logs are different for other OS versions, so we may actually be looking at multiple issues here.

Might be an environment sanitization issue. For me, this `-lphobos2` thing is erroring out way down at the "tools" building stage at the end of `install`.

Console output:
```
==> Downloading https://github.com/D-Programming-Language/tools/archive/v2.071.0.tar.gz
Already downloaded: /Library/Caches/Homebrew/dmd--tools-2.071.0.tar.gz
==> Verifying dmd--tools-2.071.0.tar.gz checksum
tar xvf /Library/Caches/Homebrew/dmd--tools-2.071.0.tar.gz
x tools-2.071.0/
...
x tools-2.071.0/tolf.d
x tools-2.071.0/update.sh
x tools-2.071.0/win32.mak
==> make install DMD=/usr/local/Cellar/dmd/2.071.0/bin/dmd INSTALL_DIR=/usr/local/Cellar/dmd/2.071.0 MODEL=64 -f posix.mak
/usr/local/Cellar/dmd/2.071.0/bin/dmd -m64 -w -ofgenerated/osx/64/rdmd rdmd.d
ld: library not found for -lphobos2
clang: error: linker command failed with exit code 1 (use -v to see invocation)
--- errorlevel 1
make: *** [generated/osx/64/rdmd] Error 1
```

From the build log, I can see this:

```
gcc called with: generated/osx/64/rdmd.o -o generated/osx/64/rdmd -m64 -Xlinker -no_compact_unwind -L/usr/local/Cellar/dmd/2.071.0/lib -lphobos2 -lpthread -lm
superenv removed:  -m64 -L/usr/local/Cellar/dmd/2.071.0/lib
superenv added:  ...
```

It looks like brew's `superenv` is scrubbing the current installation target prefix from the compiler arguments. Hmm. This seems like a legit use case for multi-step builds involving resources or auxiliary programs like this; maybe superenv should be letting that through? That `-L` seems to be the right way to reference it.

**Comment by tdsmith** [comment]
> This seems like a legit use case for multi-step builds involving resources or auxiliary programs like this; maybe superenv should be letting that through? 

That seems fine but why isn't it happening without test-bot's influence?

**Comment by apjanke** [comment]
> That seems fine but why isn't it happening without test-bot's influence?

Because `test-bot` is enabling an [alternate implementation of the `keep?` logic](https://github.com/Homebrew/brew/blob/0382134cf87321166fb8f03223e21b1d242ad7cb/Library/ENV/4.3/cc#L203-L206) in the env's tool wrappers.

```
[/usr/local/Library/Homebrew on ⇄ master]
$ grep -r HOMEBREW_EXPER *
cmd/test-bot.rb:    ENV["HOMEBREW_EXPERIMENTAL_FILTER_FLAGS_ON_DEPS"] = "1"
[/usr/local/Library/Homebrew on ⇄ master]
$
```

This will reproduce it, too.

```
[~/tmp/test-bot]
$ export HOMEBREW_EXPERIMENTAL_FILTER_FLAGS_ON_DEPS=1
[~/tmp/test-bot]
$ brew install -s dmd
...
==> make install DMD=/usr/local/Cellar/dmd/2.070.2/bin/dmd INSTALL_DIR=/usr/local/Cellar/dmd/2.070.2 MODEL=64 -f posix.m
Last 15 lines from /Users/janke/Library/Logs/Homebrew/dmd/04.make:
clang: error: linker command failed with exit code 1 (use -v to see invocation)
--- errorlevel 1
make: *** [generated/osx/64/catdoc] Error 1
ld: library not found for -lphobos2
clang: error: linker command failed with exit code 1 (use -v to see invocation)
--- errorlevel 1
```

Should be a quick fix now that we know where the issue is.

**Comment by apjanke** [comment]
I've submitted a fix at Homebrew/brew#63.

**Comment by apjanke** [comment]
I reran this after merging 63. Now it's [failing with a different error](http://bot.brew.sh/job/Homebrew%20Core%20Pull%20Requests/379/).

Looks like something related to `git` use inside the build process? Or maybe that `std.stream` deprecation warning is being treated as an error?

```
/Applications/Xcode.app/Contents/Developer/usr/bin/make -C ../druntime -f posix.mak MODEL=64 DMD=/usr/local/Cellar/dmd/2.071.0/bin/dmd OS=osx BUILD=release
fatal: Not a git repository (or any of the parent directories): .git
make[1]: Nothing to be done for `target'.
...
std/cstream.d(27): Deprecation: module std.stream is deprecated - It will be removed from Phobos in October 2016. If you still need it, go to https://github.com/DigitalMars/undeaD
std/socketstream.d(41): Deprecation: module std.stream is deprecated - It will be removed from Phobos in October 2016. If you still need it, go to https://github.com/DigitalMars/undeaD
```

I can reproduce locally on Mavericks. [Here's a build log](https://gist.github.com/c73d8d192d5e65f80d53431beb68ea45), using a single build job for readability.

**Comment by John-Colvin** [comment]
Ok, I'll take a look

File: Formula/dmd.rb (hunk 1 above)
Comment (John-Colvin) [inline Formula/dmd.rb]: that was in the commits marked `DO NOT MERGE`, I just added them to get some debug output from the auto-tester

**Comment by apjanke** [comment]
The rerun worked. Might have been a local issue on my machine.

Merged. Thank you for your contribution to Homebrew!

**Comment by John-Colvin** [comment]
The test "DO NOT MERGE" commits were causing the failure, so I removed them. Glad to have this finally sorted out :)

**Pull request closed without merging**